modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/MergeJoinPlannerTest.java

```diff
@@ -0,0 +1,2799 @@
+            "C1", Integer.class, "C2", Integer.class, "C3", Integer.class);
+
+        right.addIndex(
+            RelCollations.of(
+                new RelFieldCollation(0, DESCENDING),
+                new RelFieldCollation(1, ASCENDING),
+                new RelFieldCollation(2, ASCENDING)
+            ),
+            "idx"
+        );
+
+        IgniteSchema schema = createSchema(
+            createTable("LEFT_T", IgniteDistributions.single(), "C1", Integer.class, "C2", Integer.class, "C3", Integer.class),
+            right
+        );
+
+        String sql = "" +
+            "select * " +
+            "  from LEFT_T " +
+            "  join RIGHT_T " +
+            "    on LEFT_T.c1 = RIGHT_T.c1 " +
+            "   and LEFT_T.c2 = RIGHT_T.c2 ";
+
+        IgniteRel rel = physicalPlan(sql, schema, DISABLED_RULES);
+
+        assertNull(sortOnTopOfScan(rel, "RIGHT_T"));
+        assertEquals(
+            RelCollations.of(
+                new RelFieldCollation(0, DESCENDING),
+                new RelFieldCollation(1, ASCENDING)
+            ),
+            sortOnTopOfScan(rel, "LEFT_T").collation()
+        );
+    }
+
+    /**
+     * Test verifies the collation propagation from a parent node.
+     *
+     * In case of inner join, right collation that is superset of join keys could be derived as is.
+     *
+     * @throws Exception In case of any unexpected error.
+     */
+    @Test
+    public void testInnerDerivePreserveRight9() throws Exception {
+        TestTable right = createTable("RIGHT_T", IgniteDistributions.single(),
+            "C1", Integer.class, "C2", Integer.class, "C3", Integer.class);
+
+        right.addIndex(
+            RelCollations.of(
+                new RelFieldCollation(0, ASCENDING, RelFieldCollation.NullDirection.FIRST),
+                new RelFieldCollation(1, ASCENDING),
+                new RelFieldCollation(2, ASCENDING)
+            ),
+            "idx"
+        );
+
+        IgniteSchema schema = createSchema(
+            createTable("LEFT_T", IgniteDistributions.single(), "C1", Integer.class, "C2", Integer.class, "C3", Integer.class),
+            right
+        );
+
+        String sql = "" +
+            "select * " +
+            "  from LEFT_T " +
+            "  join RIGHT_T " +
+            "    on LEFT_T.c1 = RIGHT_T.c1 " +
+            "   and LEFT_T.c2 = RIGHT_T.c2 ";
+
+        IgniteRel rel = physicalPlan(sql, schema, DISABLED_RULES);
+
+        assertNull(sortOnTopOfScan(rel, "RIGHT_T"));
+        assertEquals(
+            RelCollations.of(
+                new RelFieldCollation(0, ASCENDING, RelFieldCollation.NullDirection.FIRST),
+                new RelFieldCollation(1, ASCENDING)
+            ),
+            sortOnTopOfScan(rel, "LEFT_T").collation()
+        );
+    }
+
+    /**
+     * Test verifies the collation propagation from a parent node.
+     *
+     * In case of inner join, right collation that is superset of join keys could be derived as is.
```

Comment:
Looks like the comment doesn't describe the test case completely.

Reply:
fixed